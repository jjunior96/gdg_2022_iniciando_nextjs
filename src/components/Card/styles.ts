import Link from 'next/link';

import { lighten } from 'polished';
import styled, { css } from 'styled-components';

export const Container = styled.div`
  ${({ theme }) => css`
    width: 100%;
    height: 32rem;

    background-color: ${theme.colors.white};

    box-shadow: ${theme.box.shadow};
    border-radius: ${theme.border.radius};

    border: 1px solid ${lighten(0.17, theme.colors.gray_200)};

    overflow: hidden;

    transition: box-shadow ${theme.transition.default},
      color ${theme.transition.default};

    &:hover {
      color: ${theme.colors.green};

      box-shadow: 0 4px 12px ${lighten(0.1, theme.colors.green)};
    }
  `}
`;

export const LinkStyled = styled(Link)``;

export const ImageContainer = styled.div`
  width: 100%;
  height: 70%;

  position: relative;

  img {
    position: absolute;

    object-fit: cover;
  }
`;

export const CardFooter = styled.footer`
  ${({ theme }) => css`
    padding: ${theme.spacings.xsmall};

    p {
      margin-top: ${theme.spacings.xxsmall};

      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
      line-height: ${theme.font.sizes.medium};
      -webkit-line-clamp: 2;
      -webkit-box-orient: vertical;
    }
  `}
`;
