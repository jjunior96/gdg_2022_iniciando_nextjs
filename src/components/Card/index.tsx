import Image from 'next/image';

import Tag from 'components/Tag';

import * as S from './styles';

export type CardProps = {
  id: number;
  name: string;
  image: string;
  status: string;
};

const Card = ({ id, name, image, status }: CardProps) => {
  return (
    <S.Container>
      <S.LinkStyled href={`/character/${id}`}>
        <a>
          <S.ImageContainer>
            <Image src={image} alt={`Thumbnail of ${name}`} layout="fill" />
          </S.ImageContainer>

          <S.CardFooter>
            <Tag color={`${status === 'Alive' ? 'green' : 'red'}`}>
              {status}
            </Tag>
            <p>{name}</p>
          </S.CardFooter>
        </a>
      </S.LinkStyled>
    </S.Container>
  );
};

export default Card;
