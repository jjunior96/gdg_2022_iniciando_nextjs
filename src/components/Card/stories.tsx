import { Story, Meta } from '@storybook/react/types-6-0';

import Card, { CardProps } from '.';

export default {
  title: 'Card',
  component: Card
} as Meta;

export const Default: Story<CardProps> = (args) => <Card {...args} />;
