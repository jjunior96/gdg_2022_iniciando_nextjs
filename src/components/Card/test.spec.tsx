import { screen } from '@testing-library/react';

import { renderWithTheme } from 'utils/tests/__config__/helpers';

import Card from '.';

const props = {
  id: 1,
  image: '',
  name: 'Card',
  status: 'Alive'
};

describe('<Card />', () => {
  it('should render the heading', () => {
    const { container } = renderWithTheme(<Card {...props} />);

    expect(screen.getByText(/Card/i)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
