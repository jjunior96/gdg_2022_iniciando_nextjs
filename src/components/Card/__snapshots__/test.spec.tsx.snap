// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> should render the heading 1`] = `
.c3 {
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  max-width: 20rem;
  padding: 0.5rem 2.4rem;
  color: #FAFAFA;
  font-size: calc(1.2rem - 0.1rem);
  font-weight: 600;
  border-style: solid;
  border-width: 1px;
  border-radius: 20px;
  overflow: hidden;
  background-color: #90e7c0;
  border-color: #24A66B;
  color: #24A66B;
}

.c0 {
  width: 100%;
  height: 32rem;
  background-color: #FAFAFA;
  box-shadow: 0 8px 8px rgba(0,0,0,0.2);
  border-radius: 0.4rem;
  border: 1px solid #c5d2d7;
  overflow: hidden;
  -webkit-transition: box-shadow 0.3s ease-in-out,color 0.3s ease-in-out;
  transition: box-shadow 0.3s ease-in-out,color 0.3s ease-in-out;
}

.c0:hover {
  color: #24A66B;
  box-shadow: 0 4px 12px #2dd086;
}

.c1 {
  width: 100%;
  height: 70%;
  position: relative;
}

.c1 img {
  position: absolute;
  object-fit: cover;
}

.c2 {
  padding: 1.6rem;
}

.c2 p {
  margin-top: 1rem;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  line-height: 2.4rem;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
}

<div
  class="c0"
>
  <a
    href="/character/1"
  >
    <div
      class="c1"
    >
      <span
        style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
      >
        <img
          alt="Thumbnail of Card"
          data-nimg="fill"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
    </div>
    <footer
      class="c2"
    >
      <div
        class="c3"
        color="green"
        role="textbox"
      >
        Alive
      </div>
      <p>
        Card
      </p>
    </footer>
  </a>
</div>
`;
